rpcperms/middleware_handler.go

```diff
@@ -0,0 +1,448 @@
+	defer h.wg.Wait()
+
+	// Create a channel that responses from middlewares are sent into.
+	responses := make(chan *lnrpc.RPCMiddlewareResponse)
+
+	// errChan is used by the receive loop to signal any errors that occur
+	// during reading from the stream. This is primarily used to shutdown
+	// the send loop in the case of an RPC client disconnecting.
+	errChan := make(chan error, 1)
+
+	// Start a goroutine to receive responses from the channel acceptor.
+	// We expect the receive function to block, so it must be run in a
+	// goroutine (otherwise we could not send more than one channel accept
+	// request to the client).
+	h.wg.Add(1)
+	go func() {
+		h.receiveResponses(errChan, responses)
+		h.wg.Done()
+	}()
+
+	return h.sendInterceptRequests(errChan, responses)
+}
+
+// receiveResponses receives responses for our channel accept requests and
+// dispatches them into the responses channel provided, sending any errors that
+// occur into the error channel provided.
+func (h *MiddlewareHandler) receiveResponses(errChan chan error,
+	responses chan *lnrpc.RPCMiddlewareResponse) {
+
+	for {
+		resp, err := h.receive()
+		if err != nil {
+			errChan <- err
+			return
+		}
+
+		select {
+		case responses <- resp:
+
+		case <-h.done:
+			return
+
+		case <-h.quit:
+			return
+		}
+	}
+}
+
+// sendInterceptRequests handles channel acceptor requests sent to us by our
+// Accept() function, dispatching them to our acceptor stream and coordinating
+// return of responses to their callers.
+func (h *MiddlewareHandler) sendInterceptRequests(errChan chan error,
+	responses chan *lnrpc.RPCMiddlewareResponse) error {
+
+	// Close the done channel to indicate that the interceptor is no longer
+	// listening and any in-progress requests should be terminated.
+	defer close(h.done)
+
+	interceptRequests := make(map[uint64]*interceptRequest)
+
+	for {
+		select {
+		// Consume requests passed to us from our Accept() function and
+		// send them into our stream.
+		case newRequest := <-h.interceptRequests:
+			id := atomic.AddUint64(&h.lastRequestID, 1)
+
+			req := newRequest.request
+			interceptRequests[id] = newRequest
+
+			interceptReq, err := req.ToRPC(id)
+			if err != nil {
+				return err
+			}
+
+			if err := h.send(interceptReq); err != nil {
+				return err
+			}
+
+		// Process newly received responses from our channel acceptor,
+		// looking the original request up in our map of requests and
+		// dispatching the response.
+		case resp := <-responses:
+			requestInfo, ok := interceptRequests[resp.RequestId]
+			if !ok {
+				continue
+			}
+
+			response := &interceptResponse{}
+			switch msg := resp.GetMiddlewareMessage().(type) {
+			case *lnrpc.RPCMiddlewareResponse_Feedback:
+				t := msg.Feedback
+				if t.Error != "" {
+					response.err = fmt.Errorf("%s", t.Error)
+					break
+				}
+
+				// For intercepted responses we also allow the
+				// content itself to be overwritten.
+				if requestInfo.request.Type == TypeResponse &&
```

Comment:
So the idea is to allow new middleware to _modify_ a normal `lnd` response? I guess a use case for this could be adding extra account ID/detail information?

Reply:
Yes exactly. The middleware is restricted to the fields the proto defines. But you can filter or add information. In our faraday accounts example we would not return any channels for example or only show the user's payments to hide some of the complexity of a node to custodial users.